### PR TITLE
Added Containers default views test

### PR DESCRIPTION
**Pull request by pavelzag**

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 6de52ae2f63d0fd19b864020382d245ab95631b1:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
1) The lint is not correct although my bot for some weird reason says nothing. Between module level functions, there should be 2 empty lines, comments do not count in them.
2) The tests need some comments. This is enforced on all new tests added. Example is [here](https://github.com/ManageIQ/integration_tests/blob/master/cfme/tests/automate/test_vmware_methods.py#L78)
3) The tests need tier marking. Something like [this](https://github.com/ManageIQ/integration_tests/blob/master/cfme/tests/automate/test_vmware_methods.py#L22). These might be tier 2.
3) Single quotes are preferred for strings. This is not enforced much, just informing :)

**Comment by mfalesni** [comment]
The comments use `"` though.

**Comment by pavelzag** [comment]
I'll edit and resend with the fixes.
Thanks for the review

**Comment by dajoRH** [comment]
Lint report for commit 4cba5f8a1fbcf0b1bb1d1b6792e425a4f3eecedc:

`cfme/tests/containers/test_containers_default_views.py`:
- :red_circle: Line 4:1: **F401** *'settings' imported but unused*
- :bangbang: Line 25:101: **E501** *line too long (160 > 100 characters)*
- :bangbang: Line 66:101: **E501** *line too long (159 > 100 characters)*
- :bangbang: Line 107:101: **E501** *line too long (157 > 100 characters)*
- :bangbang: Line 111:5: **E265** *block comment should start with '# '*
- :bangbang: Line 148:101: **E501** *line too long (156 > 100 characters)*
- :bangbang: Line 189:101: **E501** *line too long (161 > 100 characters)*
- :bangbang: Line 230:101: **E501** *line too long (162 > 100 characters)*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by pavelzag** [comment]
I cannot remove the "from cfme.configure import settings" line from the code because it would break the test. 
I am not sure why but  "sel.force_navigate('my_settings_default_views')" will not work anymore after I remove that line. There's some kind of dependency there, I am not sure why

**Comment by dajoRH** [comment]
Lint report for commit 738a716afc27d1123e776245ff6e03dec03dadd3:

`cfme/tests/containers/test_containers_default_views.py`:
- :red_circle: Line 4:1: **F401** *'settings' imported but unused*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by psav** [comment]
It's not working because the nav points are not loaded. Force navigate requires the navigation points to be loaded via that import.

**Comment by pavelzag** [comment]
Pete, can you elaborate please? Not sure I understood..

File: cfme/tests/containers/test_containers_default_views.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+# CMP-9937
+
+
+def test_containers_projects_view():
+    """
+
+    This test checks successful change of default views settings for Containers --> Projects menu
+    * Goes to Settings --> My Settings --> Default Views menu and change the default view
+    settings of Containers --> Containers Projects to Grid/Tile/List view
+    * Goes to Compute --> Containers --> Projects and verifies the selected view
```
Comment (mfalesni) [inline cfme/tests/containers/test_containers_default_views.py]: Can you please apply the same styling as in the other test descirptions?

```python
"""Test description, rather short
# an empty line in between
Steps:
    * foo the bar
    * bar the baz!
```

You can also add a Prerequisities section if that is required, same style applies.

**Comment by dajoRH** [comment]
Lint report for commit 28b3ca2086754af91bec4cbfcc06614aae0d65f9:

`cfme/tests/containers/test_containers_default_views.py`:
- :red_circle: Line 4:1: **F401** *'settings' imported but unused*
- :red_circle: Line 23:5: **F841** *local variable 'is_active' is assigned to but never used*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0260bb67898f9418585e9a381bc537b60aa83bbd:

`cfme/tests/containers/test_containers_default_views.py`:
- :red_circle: Line 4:1: **F401** *'settings' imported but unused*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@pavelzag It is very simple. We use a navigational tree for navigating around. The tree branches and leaves are defined in the files that define the UI objects, like branches for VMs are defined in eg. cfme/infrastructure/virtual_machine.py. So in order to navigate there, you need to import that module so the tree can be extended. If you use the objects, you won't face this issue because by importing the class you have imported the module which extended the naviagation tree already. But when you want to navigate somewhere without importing the respective objects, then you have to import that module manually.

In the full test run, I guess that you might not face this issue because a different test module would import that before, but that is not what we want.

We might posibly introduce some autoloading so all the "model" modules are loaded so everything is ready.

**Comment by dajoRH** [comment]
Lint report for commit 8a02f62593c03fb62fd56c921cedc27d09205e1d:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by pavelzag** [comment]
@mfalesni So if I remove the import it will work? How can I ensure that it works when I run it standalone?

**Comment by dajoRH** [comment]
Lint report for commit 0ee7fca46d4ee72e51023b89062bdc1776bf6e01:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@pavelzag The module **must** be imported. Otherwise the corresponding part of navigation tree is not in place. If you are in doubt, just import it and put a comment `# noqa` at the end of the line.

**Comment by pavelzag** [comment]
@mfalesni Ok, added the change per your request

**Comment by dajoRH** [comment]
Lint report for commit 5e93b6de122c688187a7929fccc15f6fd9ff31c2:

`cfme/tests/containers/test_containers_default_views.py`:
- :red_circle: Line 4:1: **F401** *'settings' imported but unused*
- :bangbang: Line 4:36: **E261** *at least two spaces before inline comment*
- :bangbang: Line 4:37: **E262** *inline comment should start with '# '*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e27b1e107f438d6743423da12208baeb519ea941:

`cfme/tests/containers/test_containers_default_views.py`:
- :red_circle: Line 4:1: **F401** *'settings' imported but unused*
- :bangbang: Line 4:38: **E262** *inline comment should start with '# '*

Please, rectify these issues :smirk: .
*CFME QE Bot*

**Comment by psav** [comment]
Lint needs fixing and commits need squashing

**Comment by mfalesni** [comment]
Well, I guess it might also enjoy some parametrization, I see lots of duplicate code, what is your opinion, @pavelzag  ?

**Comment by pavelzag** [comment]
@mfalesni Hi Milan, The 2 failures I see are from this line:
from cfme.configure import settings  #noqa
This is the way you wanted me to write it, no? It seems the test does require the import settings and I added #noqa in the end of the line as you you requested..

**Comment by mfalesni** [comment]
@pavelzag `E262 inline comment should start with '# '` you can notice that a comment must start with a blank line, separating the first alphanumeric character from the hash sign by one space.

I guess when you fix it, it will be OK.

If you have properly configured editor with flake8 checks, then you should see these errors automatically.

**Comment by pavelzag** [comment]
@mfalesni Ah of course.. Changed it..
When you say to add some parametrization what do you mean? 
Like adding variables outside to hold stuff like 'my_settings_default_views', 'Containers Providers' etc?

**Comment by dajoRH** [comment]
Lint report for commit 1a38fbe71a41f8b5bbf462dc0ab779c239a5ad5d:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by mfalesni** [comment]
@pavelzag Something like this what I wrote on the comment [here](https://github.com/ManageIQ/integration_tests/pull/3102#discussion_r68046086). I actually proposed two solutions in there :)

**Comment by pavelzag** [comment]
Ok, I'll try to understand this code. It's slimmer but slightly less readable for me. I'll try to rewrite it with parametrize

**Comment by mfalesni** [comment]
@pavelzag of course it isn't something you **must** use, just samples on how can you make the test so DRY that you cannot make it shorter :)

**Comment by pavelzag** [comment]
@mfalesni Oh no, I loved it. It's automatically reduced the lines of code. Will clean it up a bit and commit again later

**Comment by dajoRH** [comment]
Lint report for commit c36288fc6bb6b5b2d2240bfa6ebec23ed20ecca7:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 52295d51914257690fd604fa607ce6b580818850:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 02bfe12352a3d2e4de1dc1cfaef47653cdd3d03c:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by pavelzag** [comment]
@mfalesni Hey Milan, I updated the tests according to what you requested. Can you give them a look? Cheers!

**Comment by dajoRH** [comment]
Lint report for commit b3a5c0fffd350e3aa083cea8f6e1cb4c10e6ac4f:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 09b3f017e02dc248c255d891efbfa48b92baf989:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/tests/containers/test_containers_default_views.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+             /Containers/Replicators
+              to Grid/Tile/List view
+            * Goes to Compute --> Containers --> Providers and verifies the selected view
+
+        """
+    sel.force_navigate('my_settings_default_views')
+    bg = ButtonGroup(button_group)
+    bg.choose(view)
+    sel.click(form_buttons.save)
+    if button_group == 'Containers Providers':
```
Comment (psav) [inline cfme/tests/containers/test_containers_default_views.py]: I'd have probably created a mapping dict for this
```python
mapping = {
    'Containers Providers': 'containers_providers'
    ....
}
sel.force_navigate(mapping[button_group])
```
but I'm happy with what is here ;)

**Comment by dajoRH** [comment]
Lint report for commit 7c13a2a306ee6b008064dbd108d07cb068b52b4b:

Everything seems all right :smile: .
*CFME QE Bot*

File: cfme/tests/containers/test_containers_default_views.py (hunk 2 above)
Comment (pavelzag) [inline cfme/tests/containers/test_containers_default_views.py]: This is a great idea.. Implemented, thanks

**Comment by dajoRH** [comment]
Lint report for commit fceb7be02eb5907fb82994d0236503476bf12c78:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by pavelzag** [comment]
Hmm not sure, how do I do it?
Should I do it with the mail I am signed to GitHub (my private) or with the Red Hat one?

**Comment by pavelzag** [comment]
Hey, anybody knows the current status of this PR?

File: cfme/tests/containers/test_containers_default_views.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+from cfme.web_ui import toolbar as tb, ButtonGroup, form_buttons
+from cfme.configure import settings  # noqa
+from utils import testgen
+from utils.version import current_version
+
+pytestmark = [
+    pytest.mark.uncollectif(
+        lambda: current_version() < "5.6"),
+    pytest.mark.usefixtures('setup_provider'),
+    pytest.mark.tier(1)]
```
Comment (psav) [inline cfme/tests/containers/test_containers_default_views.py]: This shouldn't be a Tier1 test, but a 2, or maybe even 3

File: cfme/tests/containers/test_containers_default_views.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        This test checks successful change of default views settings for Containers -->
+        Providers/Projects/Routes/Nodes/Containers/Replicators menu
+        Steps:
+            * Goes to Settings --> My Settings --> Default Views menu and change the default view
+             settings of Containers --> Containers Providers/Projects/Routes/Nodes
+             /Containers/Replicators
+              to Grid/Tile/List view
+            * Goes to Compute --> Containers --> Providers and verifies the selected view
+        """
+    sel.force_navigate('my_settings_default_views')
```
Comment (psav) [inline cfme/tests/containers/test_containers_default_views.py]: Whilst I don't like this pattern, I see it is in use across the codebase, I will file an issue about it and we will rectify it later. I would prefer to see a method on an object somewhere that performs these steps. Tests shouldn't have "sel" or "clicks" in them unless absolutely necessary (ie...testing a very specific part of UI functionality)

File: cfme/tests/containers/test_containers_default_views.py (hunk 3 above)
Comment (pavelzag) [inline cfme/tests/containers/test_containers_default_views.py]: I was instructed to mark it as Tier1..

File: cfme/tests/containers/test_containers_default_views.py (hunk 4 above)
Comment (pavelzag) [inline cfme/tests/containers/test_containers_default_views.py]: How else should I navigate to the page I want to test?

File: cfme/tests/containers/test_containers_simple_search.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    """ Projects summary page - Search bar
+
+           This test checks Search bar functionality on the Projects summary page
+
+           Steps:
+               * Goes to Containers Projects page
+               * Inserts: Irregular symbol, '*' character, full search string, partial search string
+               * Verify proper results
+           """
+    if ''.join(projects_search_strings) == '*':
```
Comment (psav) [inline cfme/tests/containers/test_containers_simple_search.py]: This and the function below (at first glance) look identical, does it make sense to apply some DRY here and make this a function where "containers_projects" is an argument to the function?

**Comment by psav** [comment]
I've made a few more comments on this, I'm happy to let some pieces in, even though I would like to change them in the future, the examples we have in the code are identical, so we need to change all of them. This is a fault of ours providing a bad example. So the steps to get this merged as I see it

1) Get GPG key signed and if already done so, re push this with a signed top commit
2) Address the comments stated
3) Squash to a single commit

File: cfme/tests/containers/test_containers_default_views.py (hunk 4 above)
Comment (psav) [inline cfme/tests/containers/test_containers_default_views.py]: This comment refers to the whole function(test) not just that one line, as I mentioned don't worry about it for now, I was just marking it so I can use it in an issue I am filing.

File: cfme/tests/containers/test_containers_default_views.py (hunk 3 above)
Comment (psav) [inline cfme/tests/containers/test_containers_default_views.py]: OK

File: cfme/tests/containers/test_containers_simple_search.py (hunk 5 above)
Comment (pavelzag) [inline cfme/tests/containers/test_containers_simple_search.py]: Yes, I guess I can do that

**Comment by dajoRH** [comment]
Lint report for commit 9ac2da982f46f1982709eacdb0f81f1a4922f8d4:

Everything seems all right :smile: .
*CFME QE Bot*

**Comment by psav** [comment]
Happy to merge this, just need to see a run in PRT

**Comment by pavelzag** [comment]
What's a PRT?

**Comment by psav** [comment]
This is why we need your commit to be signed with GPG so that PRT can test it. PRT stands for Pull Request Tester and it runs your PR against multiple CFME versions.

**Comment by psav** [comment]
Moved to #3152

**Pull request closed without merging**